openstudiocore/src/openstudio_lib/ConstructionsTabController.cpp

```diff
@@ -32,29 +32,80 @@
 namespace openstudio {
 
 ConstructionsTabController::ConstructionsTabController(bool isIP, const model::Model& model)
-  : MainTabController(new ConstructionsTabView(model, "Constructions"))
+  : MainTabController(new ConstructionsTabView(model, "Constructions")),
+  m_model(model),
+  m_isIP(isIP)
 {
-  m_defaultConstructionSetsController = std::shared_ptr<DefaultConstructionSetsController>(new DefaultConstructionSetsController(model));
-  m_constructionsController = std::shared_ptr<ConstructionsController>(new ConstructionsController(isIP, model));
-  m_materialsController = std::shared_ptr<MaterialsController>(new MaterialsController(isIP, model));
+  this->mainContentWidget()->addSubTab("Construction Sets", DEFAULT_CONSTRUCTIONS);
+  this->mainContentWidget()->addSubTab("Constructions", CONSTRUCTIONS);
+  this->mainContentWidget()->addSubTab("Materials", MATERIALS);
 
-  this->mainContentWidget()->addSubTab("Construction Sets", m_defaultConstructionSetsController->subTabView(),DEFAULT_CONSTRUCTIONS);
-  this->mainContentWidget()->addSubTab("Constructions", m_constructionsController->subTabView(),CONSTRUCTIONS);
-  this->mainContentWidget()->addSubTab("Materials", m_materialsController->subTabView(),MATERIALS);
+  connect(this->mainContentWidget(), &MainTabView::tabSelected, this, &ConstructionsTabController::setSubTab);
+}
 
-  connect(this, &ConstructionsTabController::toggleUnitsClicked, static_cast<ModelSubTabView*>(m_defaultConstructionSetsController->subTabView()), &ModelSubTabView::toggleUnitsClicked);
+ConstructionsTabController::~ConstructionsTabController()
+{
+  disconnect(this->mainContentWidget(), &MainTabView::tabSelected, this, &ConstructionsTabController::setSubTab);
+}
 
-  connect(this, &ConstructionsTabController::toggleUnitsClicked, static_cast<ModelSubTabView*>(m_constructionsController->subTabView()), &ModelSubTabView::toggleUnitsClicked);
+void ConstructionsTabController::setSubTab(int index)
+{
+  if (m_currentIndex == index) {
+    return;
+  }
+  else {
+    m_currentIndex = index;
+  }
 
-  connect(this, &ConstructionsTabController::toggleUnitsClicked, static_cast<ModelSubTabView*>(m_materialsController->subTabView()), &ModelSubTabView::toggleUnitsClicked);
+  if (qobject_cast<DefaultConstructionSetsController *>(m_currentController)) {
```

Comment:
yeah ok this is all weird, but i can't argue with improvement. I would still like to understand better.